### DISCUSSION TITLE
What are strategies for moving a stream to a different set of nodes in the cluster?

### DISCUSSION THREAD

**Discussion by bruth** [nats-io/nats-server#2730] [Q&A] [answered]
The driving motivation is to deal with situations when and existing NATS cluster need to be migrated to new VMs/servers/hardware. The goal would be to not require clients to change the address they are connecting to. Obviously if client connections are temporarily interrupted and auto-reconnect, then that is fine (no different then a network hiccup).

To setup a scenario, let's assume I have a 3-node cluster with Jetstream enabled on all nodes. I have one or more streams some of which are configured to have 3 replicas. The problem is that I am running out of storage and need to add more storage for existing streams but also new streams.

I am aware that additional nodes can be added to a cluster with JS enabled and NATS will distribute stream replicas across the set of available nodes. My understanding is that replication/quorum happens on a per stream basis.

Would it be possible to increase the size of the cluster, with the new nodes having the new storage requirements, and the stream be moved transparently to these new nodes? I know a mirror could be setup, but this would require clients to know to switch over to publishing and consuming to this new stream.

Is there some other transparent way of handling this?

**Comment by ripienaar** [comment]
Generally you'd want to keep your server names static, so in your 3 node cluster case you'd bring up replacement nodes on the same name - possibly CNAME of course - and with appropriate TLS certs etc.

When you cleanly shut down a old one and bring a new one it would start syncing data soon as it comes online of any streams assigned to it.  You want to wait for this process to fully complete, all streams in all accounts to be up to date - then move to the next machine.  Always use Lame Duck Mode to prepare a server for replacement.

Otoh if you increase size and you want to redistribute streams, you'll have to use the cli to remove one peer from a stream and it will select a new one at random

Both processes quite involved and manual and requires careful testing and planning.

**Reply by bruth** [reply]
Thanks. To confirm the strategy:

1. Add the new desired nodes to the cluster
2. Take one of the existing nodes offline
3. Explicitly remove the node as a peer `nats cluster peer-remove <stream> <peer>`
   - This will cause JS to pick one of the new nodes at random and replicate the data to it
4. Repeat 2-3 until all existing nodes are gone

I understand this is tricky since there could be multiple streams and needing to wait/ensure replication is complete.

**Reply by ripienaar** [reply]
Yes, if adding new nodes isn’t a problem (names changing etc)

replacing them one by one is the other option

**Comment by derekcollison** [comment]
Depending on your needs you can also backup and restore into a new cluster with different storage properties.

**Reply by bruth** [reply]
Thanks. This is indeed the simplest approach if a bit of downtime is allowed. I am not ruling this out for my purpose, but I was curious about the *no downtime* approach.

**Comment by derekcollison** [comment]
For no downtime, I would expand the cluster (JetStream assets live within a cluster scope, so any option of no downtime needs to work inside that structure) with additional nodes running JS. I would then move the non-leader peers for each asset that needs to be moved onto the new servers. This may be tricky since right now if you remove a peer it will pick a new one but _could_ pick one of the old ones. Will think more on that.

Then move the leader by asking the asset group to elect a new leader and then remove that peer after that operation has been successful.

**Reply by bruth** [reply]
> This may be tricky since right now if you remove a peer it will pick a new one but could pick one of the old ones.

It could pick an old, but presuming I take the old one offline, it would always pick a new one since two replicas of a stream can't reside on the same node, correct?

**Reply by derekcollison** [reply]
Yes but to have no interrupts you can't block the asset meaning you will have to have some old assets running for business continuity.

**Reply by bruth** [reply]
Ah right, understood thanks. I am going to try testing this approach and will document the steps.

**Reply by bruth** [reply]
Put together this 11 min video showing a quick demo of the approach. https://www.youtube.com/watch?v=uj0-Wy0Td3Y I will create a repo with the code soon. Any feedback/suggestions are welcome.

**Reply by derekcollison** [reply]
Nice will take a look..

**Reply by ripienaar** [reply]
Nice. 

Instead of peer remove on each stream you can lame duck the departing server and then remove that server from the cluster with the “server raft” command. 

This would move all data from all streams elsewhere. 

Of course though if you have lots of data you might want to do them per stream to spread the period of replication as that’s quite a heavy process. 

I want to support something like marking a node as being not a possible scheduling target so you can mark your departing nodes as not candidates for putting new streams to minimise this Random luck situation. 

Regarding the server names. If you start with a cluster of 3 and add 3 new nodes the servers will gossip that topology change to connected clients. So your clients would already know about these new nodes for the purpose of reconnection. Of course new connection you have issues

So an alternative approach would be:

start with 3
add 2 more nodes 
replace your first 3 one at a time with same names new machines 
once those 3 are all on new hardware remove the extra 2 one by one. 

Your end state is 3 machines on tbe same names as before and a same sized cluster. No client changes needed. The extra 2 temporary nodes gave you a safety net keeping things reliable during the rebuild phase. Avoidable if you are able to fast rebuild machines but if it’s a bit longer this can give you some peace of mind while avoiding load balancers.

**Reply by bruth** [reply]
Awesome thanks for the feedback, I will try out the suggestions.

**Reply by bruth** [reply]
New ~6m with the incorporated changes: https://www.youtube.com/watch?v=3HhvgV9WEYI

**Reply by ripienaar** [reply]
Nice, this seemed a bit rougher going than the first attempt (saw some n2! near the end there), when cycling machines you have to be careful that all the streams thats being removed from a machine you're retiring has been fully synced onto new machines.

Imagine you have a R3 on nodes 1,2,3.  You take out 1 and lets say all goes to 4 instead.  If 2 and 3 have not finished syncing up 4 completely 100% over all streams in all accounts and you move to take out 2, you can end up with data loss.  So you want to be super careful around how you decide to move to the next server in the loop.

**Reply by bruth** [reply]
Yea, the rougher going was due to cycling the nodes rather than the streams. Intuitively (and as you said in the previous comment), moving a single stream at a time should be _simpler_ than a full node where it can disrupt multiple streams.

For the purpose of the demo, I shortened the cycle time, but in practice it could use some signal of replication status before doing the next node cycle. Ideally not dependent on time, but health of all stream replicas on the node being cycled.

**Reply by bruth** [reply]
My next effort is going to be extracting some stats (replica lag, active replicas, etc) across all streams and consumers as it applies to a node being removed before moving onto the next node to migrate.

**Reply by bruth** [reply]
I wrote a little health summary to visualize the replication lag of the node that is in LDM. I am not sure what other metrics I would add in to model the health of a stream or consumer.

<img width="408" alt="Screen Shot 2021-12-12 at 7 42 37 AM" src="https://user-images.githubusercontent.com/515142/145712710-2cc46b02-dcc3-457a-b6f5-571955ba8ffc.png">

One thing I did observe is that even if a new node is available to be replicated onto, the server doesn't proactively start replicating to a new node when another node goes into LDM. It seems like it's only when the LDM node is removed. Is this by design?

I suppose there is the use case for putting a server in LDM to restart it and then its available again as a peer in the cluster? If so, I wonder if there could be a LDM option indicating the eventual removal of the node, so replicas should start moving right away to other nodes if available.

**Comment by ripienaar** [comment]
LDM gracefully shuts a node down moving its leadships elsewhere etc. you should always shut down a JS node that way if possible. 

but it might still come, maybe just maintenance or something. It’s not till you remove it from the RAFT group that a new member is picked in it’s place

**Reply by bruth** [reply]
Thanks, that I understand. I guess I was just suggesting some kind of joint "LDM + intention of RAFT group removal" signal. So the server knows that the node _will_ be removed or it can even remove it automatically once all clients disconnect and replicas have been moved to other nodes.

**Comment by bruth** [comment]
Created a repo with the progress thus far: https://github.com/bruth/nats-zero-down/tree/main/replace-cluster-nodes. Further work will be updated there.

**Comment by jnmoyne** [comment, marked as answer]
This question is pretty old but in the interest of cleanup: you can now control the placement of streams within a cluster or even between clusters using the [placement tags](https://docs.nats.io/nats-concepts/jetstream/streams#placement).

You can even move existing streams using those placements tags, without any interruption of service, and in any case the client applications do not need to know where a particular stream is located it's all completely transparent to the client applications.
